### PR TITLE
Allow kernel extensions to depend on other kernel extensions.

### PR DESCRIPTION
**Pull request by jakewins**
- Introduce a simple algorithm for deducing the required instantiation-order
  of kernel extensions by looping over factories over and over, instantiating each one we
  have all dependencies for, until none remain.

  Not super efficient, but since kernel extensions are relatively few and only
  instantiated once, at boot time, this shouldn't be an issue any time soon.

  This will allow us to move to a more modular approach over time, with major
  components declaring which components they require, and each properly
  implementing lifecycle.

  It does move further in a direction that risks obfuscating the dependency
  tree - but frankly the dependency tree is hard to follow anyway, and this way
  it will at least be explicitly expressed in the KernelExtensionFactory
  classes.

  This is desired, although not required, by Mjolnir, which wishes to consist
  of multiple services with clean interfaces between each other and other
  database services. Rather than introducing more manually written component
  wiring code, this solves it in a general way for services in the kernel.

### CODE REVIEW

**Comment by jexp** [comment]
Does this change KernelExtension APIs so that implemented extensions will stop working? Or is it an optional dependency declaration? Didn't look at the code tbh.

**Comment by thobe** [comment]
@jexp from reading the code it does not look like it does. But on the other hand that does not matter since `KernelExtension` is an internal API, since we control all implementations we could simply update them all should that have been the case.
